Client/Frontend/Browser/TopTabsViewController.swift

```diff
@@ -0,0 +1,340 @@
+        }
+        collectionView.snp_makeConstraints { make in
+            make.top.bottom.equalTo(view)
+            make.leading.equalTo(privateTab.snp_trailing).offset(-TopTabsUX.CollectionViewPadding)
+            make.trailing.equalTo(tabsButton.snp_leading).offset(TopTabsUX.CollectionViewPadding)
+        }
+        
+        view.backgroundColor = UIColor.blackColor()
+        updateTabCount(tabsToDisplay.count)
+        tabsButton.applyTheme(Theme.NormalMode)
+    }
+    
+    func updateTabCount(count: Int, animated: Bool = true) {
+        self.tabsButton.updateTabCount(count, animated: animated)
+    }
+    
+    func tabsTapped() {
+        delegate?.topTabsDidPressTabs()
+    }
+    
+    override func viewDidAppear(animated: Bool) {
+        super.viewDidAppear(animated)
+        collectionView.reloadData()
+        self.scrollToCurrentTab(false)
+    }
+    
+    func newTabTapped() {
+        if let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab),
+            let cell  = collectionView.cellForItemAtIndexPath(NSIndexPath(forItem: index, inSection: 0)) as? TopTabCell {
+            cell.selectedTab = false
+            if index > 0 {
+                cell.seperatorLine = true
+            }
+        }
+        delegate?.topTabsDidPressNewTab()
+        collectionView.performBatchUpdates({ _ in
+            let count = self.collectionView.numberOfItemsInSection(0)
+            self.collectionView.insertItemsAtIndexPaths([NSIndexPath(forItem: count, inSection: 0)])
+            }, completion: { finished in
+                if finished {
+                    self.scrollToCurrentTab()
+                }
+        })
+    }
+    
+    func privateTabTapped() {
+        delegate?.topTabsDidPressPrivateTab()
+        self.collectionView.reloadData()
+        scrollToCurrentTab(false)
+    }
+    
+    func closeTab() {
+        delegate?.topTabsDidPressTabs()
+    }
+    
+    func reloadFavicons() {
+        self.collectionView.reloadData()
+    }
+    
+    func scrollToCurrentTab(animated: Bool = true) {
+        guard let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab) where !collectionView.frame.isEmpty else {
+            return
+        }
+        if let frame = collectionView.layoutAttributesForItemAtIndexPath(NSIndexPath(forRow: index, inSection: 0))?.frame {
+            // Padding is added to ensure the tab is completely visible (none of the tab is under the fader)
+            let padFrame = frame.insetBy(dx: -(TopTabsUX.TopTabsBackgroundShadowWidth+TopTabsUX.FaderPading), dy: 0)
+            collectionView.scrollRectToVisible(padFrame, animated: true)
+        }
+    }
+    
+    enum TopTabsCurveDirection {
+        case Right
+        case Left
+        case Both
+    }
+    
+    static func TopTabsCurve(width: CGFloat, height: CGFloat, direction: TopTabsCurveDirection) -> UIBezierPath {
```

Comment:
Nit: Now that we have a separate file for the views, I would move this and the enum to that file instead of the VC. Also, why not put this as an extension on `UIBezierPath`?